src/ray/raylet/node_manager.cc

```diff
@@ -563,50 +552,46 @@ void NodeManager::SpillObjects(const std::vector<ObjectID> &objects_ids_to_spill
           if (!status.ok()) {
             RAY_LOG(ERROR) << "Failed to send object spilling request: "
                            << status.ToString();
+            if (callback) {
+              callback(status);
+            }
           } else {
-            RAY_CHECK(static_cast<size_t>(r.spilled_objects_url_size()) ==
-                      objects_ids.size());
             for (size_t i = 0; i < objects_ids.size(); ++i) {
               const ObjectID &object_id = objects_ids[i];
               const std::string &object_url = r.spilled_objects_url(i);
               RAY_LOG(DEBUG) << "Object " << object_id << " spilled at " << object_url;
-              // TODO(suquark): write to object directory.
-              spilled_objects_[object_id] = object_url;
-              auto search = pinned_objects_.find(object_id);
-              if (search != pinned_objects_.end()) {
-                pinned_objects_.erase(search);
-              } else {
-                RAY_LOG(ERROR) << "The spilled object " << object_id.Hex()
-                               << " is not pinned.";
-              }
+              // Write to object directory. Wait for the write to finish before
+              // releasing the object to make sure that the spilled object can
+              // be retrieved by other raylets.
+              RAY_CHECK_OK(gcs_client_->Objects().AsyncAddSpilledUrl(
+                  object_id, object_url, [this, object_id, callback](Status status) {
+                    RAY_CHECK_OK(status);
+                    // Unpin the object.
+                    // NOTE(swang): Due to a race condition, the object may not be in
+                    // the map yet. In that case, the owner will respond to the
+                    // WaitForObjectEvictionRequest and we will unpin the object
+                    // then.
+                    pinned_objects_.erase(object_id);
+                    if (callback) {
+                      callback(status);
+                    }
+                  }));
             }
           }
-          if (callback) {
-            callback(status);
-          }
         });
   });
 }
 
-void NodeManager::RestoreSpilledObjects(
-    const std::vector<ObjectID> &object_ids,
+void NodeManager::AsyncRestoreSpilledObject(
```

Comment:
Is there any way we can refactor this code to be unit testable?

Reply:
I think the best way would be to refactor the object manager code for that to be unit testable, but that's a significant effort that I don't think we should do here...

Reply:
By the way, I would have moved this into the object manager, which would've been ideal for unit-testing, but unfortunately this needs access to the worker pool, which is private to the node manager. We can think about how to pull it out later.

Reply:
Can't we just pass the shared pointer or something?